### PR TITLE
refactor player prefs

### PR DESCRIPTION
**Pull request by joelhooks**
this stabilizes video player preferences and makes them more manageable.

### CODE REVIEW

File: src/components/EggheadPlayer/index.js [hunk 1]
```diff
@@ -109,23 +103,78 @@ export default class EggheadPlayer extends Component {
...
     const active = Player.canPlay(url)
-    const {...activeProps} = this.props
+    const {resource, ...activeProps} = this.props
     const props = active ? {...activeProps, ref: this.ref} : {}
     const {
-      onPlaybackRateChange,
       playbackRate,
-      onVolumeChange,
       volumeRate,
-    } = useEggheadPlayer()
```
Comment (joelhooks) [inline src/components/EggheadPlayer/index.js:119]: the `use` prefix is very specifically for **react hooks** and hooks can't be used in old school class components like this one!

File: src/components/EggheadPlayer/index.js [hunk 2]
```diff
@@ -109,23 +103,78 @@ export default class EggheadPlayer extends Component {
...
+      videoQuality,
+      subtitle,
+      muted,
+      autoplay,
+    } = getPlayerPrefs()
```
Comment (joelhooks) [inline src/components/EggheadPlayer/index.js:115]: import `getPlayerPrefs` which simply loads the cookie which doesn't need to be a hook

File: src/components/EggheadPlayer/index.js [hunk 3]
```diff
@@ -109,23 +103,78 @@ export default class EggheadPlayer extends Component {
...
+    const onVideoQualityChanged = ({targetQuality}) => {
+      const videoQuality = {
+        id: targetQuality.id,
+        bitrate: targetQuality.bitrate,
+        label: targetQuality.label,
+        width: targetQuality.width,
+        height: targetQuality.height,
+      }
+
+      savePlayerPrefs({videoQuality})
```
Comment (joelhooks) [inline src/components/EggheadPlayer/index.js:128]: we can just save the option we want to update directly

File: src/components/EggheadPlayer/index.js [hunk 4]
```diff
@@ -109,23 +103,78 @@ export default class EggheadPlayer extends Component {
...
+
+      track(`set video quality`, {
```
Comment (joelhooks) [inline src/components/EggheadPlayer/index.js:130]: these are also good to track!

File: src/components/EggheadPlayer/players/Bitmovin.js [hunk 5]
```diff
@@ -4,6 +4,9 @@ import get from 'lodash/get'
 
 import Base from './Base'
 
+import {track} from 'utils/analytics'
+import {savePlayerPrefs} from '../use-egghead-player'
```
Comment (joelhooks) [inline src/components/EggheadPlayer/players/Bitmovin.js:8]: we can just save prefs here in the player too, which is super nice

File: src/components/EggheadPlayer/players/Bitmovin.js [hunk 6]
```diff
@@ -105,11 +113,26 @@ export default class Bitmovin extends Base {
...
         playbackspeedchanged: (e) => {
+          track('set played speed', {
+            speed: e.to,
+          })
           onPlaybackRateChange(e.to)
         },
         volumechanged: (e) => {
           onVolumeChange(e.targetVolume)
         },
+        stallstarted: () => {
```
Comment (joelhooks) [inline src/components/EggheadPlayer/players/Bitmovin.js:124]: https://cdn.bitmovin.com/player/web/8/docs/enums/core_events.playerevent.html

this makes is really easy to do event stuff without changing a bunch of files if we don't need to

File: src/components/EggheadPlayer/props.js [hunk 7]
```diff
@@ -30,6 +32,7 @@ export const propTypes = {
   onEnded: func,
   onError: func,
   onDuration: func,
+  onMuted: func,
```
Comment (joelhooks) [inline src/components/EggheadPlayer/props.js:35]: added this before I realized you could just listen for events in the player config which is way easier imo

File: src/components/EggheadPlayer/use-egghead-player.ts [hunk 8]
```diff
@@ -212,44 +213,69 @@ const onEnded = (lesson: {
...
 function onError() {}
 
-function onVideoQualityChanged() {}
-
-function onSubtitleChange() {}
-
-const getPlayerPrefs = () => cookies.get('egghead-player-prefs')
+const PLAY_PREFS_KEY = 'egghead-player-prefs'
+
+const defaultPlayerPreferences = {
```
Comment (joelhooks) [inline src/components/EggheadPlayer/use-egghead-player.ts:222]: sensible defaults!

File: src/components/EggheadPlayer/use-egghead-player.ts [hunk 9]
```diff
@@ -212,44 +213,69 @@ const onEnded = (lesson: {
...
+    kind: null,
+    label: 'off',
+    lang: null,
+  },
+  muted: false,
+}
 
-function onPlaybackRateChange(playbackRate: number) {
-  cookies.set('egghead-player-prefs', {...getPlayerPrefs(), playbackRate})
+export const getPlayerPrefs = () => {
```
Comment (joelhooks) [inline src/components/EggheadPlayer/use-egghead-player.ts:241]: the `getPlayerPrefs` looks for the cookie and if it isn't there sets the cookie with our defaults 🏆

File: src/components/EggheadPlayer/use-egghead-player.ts [hunk 10]
```diff
@@ -212,44 +213,69 @@ const onEnded = (lesson: {
...
 }
 
-function onVolumeChange(volumeRate: number) {
-  cookies.set('egghead-player-prefs', {...getPlayerPrefs(), volumeRate})
+export const savePlayerPrefs = (options: any) => {
+  console.log(`storing preference`, options)
+  return cookies.set(PLAY_PREFS_KEY, {
+    ...defaultPlayerPreferences,
+    ...getPlayerPrefs(),
+    ...options,
```
Comment (joelhooks) [inline src/components/EggheadPlayer/use-egghead-player.ts:253]: we pass in an `options` object like `{muted: true}` and it saves the whole prefs

File: src/components/EggheadPlayer/use-egghead-player.ts [hunk 11]
```diff
@@ -212,44 +213,69 @@ const onEnded = (lesson: {
...
+  })
 }
 
-export default function useEggheadPlayer(lesson: any) {
-  const playbackRate = getPlayerPrefs()?.playbackRate
-  const volumeRate = getPlayerPrefs()?.volumeRate
+export default function useEggheadPlayer(lesson: LessonResource) {
```
Comment (joelhooks) [inline src/components/EggheadPlayer/use-egghead-player.ts:257]: made this into an actual Hook, but not I don't think it's being used...

File: src/lib/lessons.ts [hunk 12]
```diff
@@ -39,6 +39,7 @@ export async function loadLesson(slug: string, token?: string) {
         free_forever
         http_url
         media_url
+        lesson_view_url
```
Comment (joelhooks) [inline src/lib/lessons.ts:42]: we haven't been watching progress because this url is missing 😱

File: src/pages/lessons/[slug].tsx [hunk 13]
```diff
@@ -70,22 +72,31 @@ const VIDEO_MIN_HEIGHT = 480
 
 const Lesson: FunctionComponent<LessonProps> = ({initialLesson}) => {
   const {height} = useWindowSize()
-
+  const [lesson, setLesson] = React.useState<any>(initialLesson)
```
Comment (joelhooks) [inline src/pages/lessons/[slug].tsx:75]: lesson is state

File: src/pages/lessons/[slug].tsx [hunk 14]
```diff
@@ -70,22 +72,31 @@ const VIDEO_MIN_HEIGHT = 480
...
+  const {slug} = lesson
   const clientHeight = isBrowser() ? height : 0
   const [lessonMaxWidth, setLessonMaxWidth] = useState(0)
   const router = useRouter()
   const playerRef = React.useRef(null)
   const {viewer} = useViewer()
   const [playerState, send] = useMachine(playerMachine)
+  const {onProgress, onEnded} = useEggheadPlayer(lesson)
```
Comment (joelhooks) [inline src/pages/lessons/[slug].tsx:83]: these are our callbacks so we can track progress

File: src/pages/lessons/[slug].tsx [hunk 15]
```diff
@@ -70,22 +72,31 @@ const VIDEO_MIN_HEIGHT = 480
...
 
-  const lesson: any = {...initialLesson}
-
   useLastResource(lesson)
 
   const {data} = useSWR(lesson.media_url, fetcher)
 
+  React.useEffect(() => {
+    const token = getAccessTokenFromCookie()
+    loadLesson(slug, token).then((loadedLesson) => {
```
Comment (joelhooks) [inline src/pages/lessons/[slug].tsx:93]: we want to load the full lesson with the viewer's credentials and merge that into our stateful lesson

**Review by DrShpongle — APPROVED** [review]
Cool! Thanks for the comments 👍

**Pull request merged**